### PR TITLE
Graphical view should open like a side preview, just like Markdown files

### LINKED ISSUE #17: Graphical view should open like a side preview, just like Markdown files
When editing markdown files there's an "Open preview to the side button" that opens a rendered view of the document on a split window. Our plugin should do the same. It would be much more discoverable than the current "right click on the right file" approach.

### LINKED ISSUE #23: Refactor ComponentViewerProvider to support clean uri
Use uri such as viewer://xcomponent/document_name

### PR DESCRIPTION
**Pull request by nicolaserny**
fixes #17 
fixes #23

### CODE REVIEW

File: extension/package.json [hunk 1]
```diff
@@ -33,14 +33,15 @@
   "contributes": {
     "commands": [
       {
-        "command": "component-viewer",
-        "title": "XCViewer: View cxml file"
+        "command": "xcomponent.preview",
+        "title": "XComponent: preview component",
```
Comment (maurelio1234) [inline extension/package.json]: component preview?

File: extension/src/componentModelProvider.ts [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+        const rawModel = this.editor.document.getText();
+
+        return {
+            model: rawModel,
+            graphical: this.getGraphicalRawModel()
+        };
+    }
+
+    private getGraphicalRawModel(): string | undefined {
+        const graphicalFile = `${this.editor.document.fileName.replace(cxmlExtension, "")}${graphicalFileSuffix}`;
```
Comment (maurelio1234) [inline extension/src/componentModelProvider.ts]: use `path.parse` in combination with `path.format` to make sure the code does exactly what we want it to do.

File: extension/test/extension.componentModelProvider.test.ts [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+        });
+    });
+
+    const testIsCxmlDocumentTests = [
+        { input: cxmlFileName, expect: true },
+        { input: otherInputFileName, expect: false }
+    ];
+
+    testIsCxmlDocumentTests.forEach(test => {
+        it(`Given ${test.input} as input file, isCxml should return ${test.expect}`, () => {
```
Comment (maurelio1234) [inline extension/test/extension.componentModelProvider.test.ts]: isCxmlDocument

File: extension/test/extension.componentModelProvider.test.ts [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+            return vscode.workspace.openTextDocument(path.join(inputPath, test.input)).then(document => {
+                return vscode.window.showTextDocument(document);
+            }).then(editor => {
+                let componentModelProvider = new ComponentModelProvider();
+                componentModelProvider.isCxmlDocument().should.equal(test.expect);
+            }, reason => {
+                should.fail(reason, undefined, "Raw model provider failed", "");
```
Comment (maurelio1234) [inline extension/test/extension.componentModelProvider.test.ts]: can we be more precise in the error message?
if not remove it...

File: extension/test/extension.componentModelProvider.test.ts [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+        });
+    });
+
+    it("If no document is opened, isCxml should return false", () => {
+        return vscode.commands.executeCommand("workbench.action.closeAllEditors")
+            .then(() => {
+                let componentModelProvider = new ComponentModelProvider();
+                componentModelProvider.isCxmlDocument().should.be.false;
+            }, reason => {
+                should.fail(reason, undefined, "Raw model provider failed", "");
```
Comment (maurelio1234) [inline extension/test/extension.componentModelProvider.test.ts]: same

File: extension/test/extension.componentViewerProvider.test.ts [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+        it(`Given ${test.input} as input file, should display ${test.expected} as html output`, () => {
+            const expectedHtml = fs.readFileSync(path.join(outputPath, expectedErrorHtmlFileName)).toString();
+            return vscode.workspace.openTextDocument(path.join(inputPath, otherInputFileName)).then(document => {
+                return vscode.window.showTextDocument(document);
+            }).then(editor => {
+                const extensionContextMock: TypeMoq.IMock<vscode.ExtensionContext> = TypeMoq.Mock.ofType<vscode.ExtensionContext>();
+                extensionContextMock.setup(x => x.extensionPath).returns(() => "/test/extensionpath");
+                const provider = new ComponentViewerProvider(extensionContextMock.object);
+                const htmlPreview = provider.provideTextDocumentContent(null);
+                htmlPreview.should.equal(expectedHtml);
```
Comment (maurelio1234) [inline extension/test/extension.componentViewerProvider.test.ts]: I personally prefer a test like `htmlPreview.contains("Cannot preview file")`, 
it has a much smaller cost of maintenance

File: extension/test/index.ts [hunk 7]
```diff
@@ -15,7 +15,7 @@ let testRunner = require("vscode/lib/testrunner");
 // You can directly control Mocha options by uncommenting the following lines
 // See https://github.com/mochajs/mocha/wiki/Using-mocha-programmatically#set-options for more info
 testRunner.configure({
-    ui: "tdd", 		// the TDD UI is being used in extension.test.ts (suite, test, etc.)
+    ui: "bdd", 		// the TDD UI is being used in extension.test.ts (suite, test, etc.)
```
Comment (maurelio1234) [inline extension/test/index.ts]: change or remove comment?

File: extension/package.json (hunk 1 above)
Comment (nicolaserny) [inline extension/package.json]: it is the name of our extension command.
It's better to preview our commands with the name of our extension.
It seems to be a best practice...

File: extension/src/componentModelProvider.ts (hunk 2 above)
Comment (nicolaserny) [inline extension/src/componentModelProvider.ts]: +1 I do it right now

File: extension/package.json (hunk 1 above)
Comment (nicolaserny) [inline extension/package.json]: Done

File: extension/src/componentModelProvider.ts (hunk 2 above)
Comment (nicolaserny) [inline extension/src/componentModelProvider.ts]: Done

File: extension/test/extension.componentModelProvider.test.ts (hunk 3 above)
Comment (nicolaserny) [inline extension/test/extension.componentModelProvider.test.ts]: Done

File: extension/test/extension.componentModelProvider.test.ts (hunk 4 above)
Comment (nicolaserny) [inline extension/test/extension.componentModelProvider.test.ts]: Done - removed unnecessary message

File: extension/test/extension.componentViewerProvider.test.ts (hunk 6 above)
Comment (nicolaserny) [inline extension/test/extension.componentViewerProvider.test.ts]: Nothing done to avoid breaking parametrised tests

**Pull request merged**